packages/core/src/DiscoveryCache.php

```diff
@@ -32,24 +33,40 @@ public function __construct(
         );
     }
 
-    public function restore(string $className): ?DiscoveryItems
+    /**
+     * @return array<class-string<\Tempest\Discovery\Discovery>, DiscoveryItems>
+     */
+    public function restore(DiscoveryLocation $location): ?array
```

Comment:
The core idea of this PR is that we cache discovery based on location instead of based on discovery class. Switching this around means we only have to loop twice over the while filesystem, instead of doing it once per discovery class.